File: test/webui/permission_queue.test.js
'use strict';

// More than one permission request can be outstanding at once — an
// Orchestrate fanout raises up to four — and the broker keeps one
// question per waiter rather than one flag for the session. The page
// must show the oldest unresolved request, show the next when one
// resolves, and unlock the composer only when the queue empties.
//
// Two failures this guards: answering the second of two requests closed
// the first without answering it, stranding its turn; and on log replay
// a request answered days ago closed a live modal, because any
// permission.resolved closed the modal regardless of its id.

const test = require('node:test');
const assert = require('node:assert/strict');

const { load } = require('./harness');

function ask(app, seq, id, description) {
  app.sse.emit({
    seq,
    type: 'permission.request',
    data: { id, tool: 'bash', description: description || id, can_always: true, rule: 'bash(*)' },
  });
}

function resolve(app, seq, id) {
  app.sse.emit({ seq, type: 'permission.resolved', data: { id, allow: true, scope: 'once' } });
}

const modalText = (app) => app.el('permission-text').textContent;
const modalOpen = (app) => app.permissionRequest.isOpen;
// Spread into this realm first: the queue lives in the harness's VM
// context, and assert.deepStrictEqual treats a cross-realm array as a
// different structure even when its contents match.
const queueIDs = (app) => [...app.state.pendingPermissionQueue].map((q) => q.id);

test('a second request waits behind the first instead of replacing it', async () => {
  const app = await load();
  ask(app, 1, 'p1', 'first command');
  ask(app, 2, 'p2', 'second command');

  assert.equal(app.state.pendingPermissionID, 'p1');
  assert.ok(modalText(app).includes('first command'), modalText(app));
  assert.deepEqual(queueIDs(app), ['p2']);
  assert.equal(modalOpen(app), true);
});

test('resolving the second request leaves the first on screen', async () => {
  const app = await load();
  ask(app, 1, 'p1', 'first command');
  ask(app, 2, 'p2', 'second command');
  resolve(app, 3, 'p2');

  assert.equal(app.state.pendingPermissionID, 'p1');
  assert.ok(modalText(app).includes('first command'), modalText(app));
  assert.deepEqual(queueIDs(app), []);
  assert.equal(modalOpen(app), true);
  assert.equal(app.el('input').disabled, true);
});

test('resolving the shown request promotes the next one', async () => {
  const app = await load();
  ask(app, 1, 'p1', 'first command');
  ask(app, 2, 'p2', 'second command');
  resolve(app, 3, 'p1');

  assert.equal(app.state.pendingPermissionID, 'p2');
  assert.ok(modalText(app).includes('second command'), modalText(app));
  assert.equal(modalOpen(app), true);
  assert.equal(app.el('input').disabled, true);
});

test('a resolution for another id leaves a live modal alone', async () => {
  const app = await load();
  ask(app, 1, 'p1', 'first command');
  resolve(app, 2, 'p9');

  assert.equal(app.state.pendingPermissionID, 'p1');
  assert.equal(modalOpen(app), true);
  assert.equal(app.el('input').disabled, true);
});

// Both halves of every permission live in the log, so opening a session
// replays each old answer in turn. Without the drop, answering nothing
// fills the queue with answered questions from last week.
test('replay drops answered pairs and shows the live request', async () => {
  const app = await load();
  ask(app, 1, 'p-old-1');
  resolve(app, 2, 'p-old-1');
  ask(app, 3, 'p-old-2');
  resolve(app, 4, 'p-old-2');
  ask(app, 5, 'p-live', 'live command');

  assert.equal(app.state.pendingPermissionID, 'p-live');
  assert.deepEqual(queueIDs(app), []);
  assert.ok(modalText(app).includes('live command'), modalText(app));
});

test('the composer unlocks only when the queue empties', async () => {
  const app = await load();
  ask(app, 1, 'p1', 'first command');
  ask(app, 2, 'p2', 'second command');

  app.el('permission-allow').click();
  await app.settle();
  assert.equal(app.callsTo('POST', '/api/sessions/sess-1/permissions/p1').length, 1);
  assert.equal(modalOpen(app), true, 'the second request should have come up');
  assert.ok(modalText(app).includes('second command'), modalText(app));
  assert.equal(app.el('input').disabled, true, 'the composer unlocked with a question still on screen');

  app.el('permission-allow').click();
  await app.settle();
  assert.equal(app.callsTo('POST', '/api/sessions/sess-1/permissions/p2').length, 1);
  assert.equal(modalOpen(app), false);
  assert.equal(app.el('input').disabled, false);
});

test('cancelling with two outstanding shows the second', async () => {
  const app = await load();
  ask(app, 1, 'p1', 'first command');
  ask(app, 2, 'p2', 'second command');

  app.sse.emit({ seq: 3, type: 'turn.cancelled', data: {} });
  // The daemon answers the cancelled question on its way out; that
  // resolution must not take the promoted one down with it.
  resolve(app, 4, 'p1');

  assert.equal(app.state.pendingPermissionID, 'p2');
  assert.ok(modalText(app).includes('second command'), modalText(app));
  assert.equal(modalOpen(app), true);
});

File: internal/tui/permission_queue_test.go
package tui

import (
	"testing"

	"localcode/internal/events"
)

// An Orchestrate fanout raises up to four permission requests at once,
// and the broker holds one question per waiter rather than one flag for
// the session. The screen must show the oldest unresolved request, show
// the next when one resolves, and unlock the composer only when the
// queue empties — and replaying the log must drop each answered request
// instead of stacking modals from last week behind the live one.
func TestPermissionQueueKeepsTheOldestUnresolvedRequestOnScreen(t *testing.T) {
	req := func(id string) events.Event {
		return events.Event{Type: events.TypePermissionRequest, Data: map[string]any{
			"id": id, "tool": "bash", "description": "run " + id, "rule": "run *",
		}}
	}
	res := func(id string) events.Event {
		return events.Event{Type: events.TypePermissionResolved, Data: map[string]any{
			"id": id, "allow": true, "scope": "once",
		}}
	}
	cases := []struct {
		name string
		evs  []events.Event
		// wantShown is the id on screen, or "" for no modal.
		wantShown string
		// wantQueued are the ids waiting behind it, oldest first.
		wantQueued []string
	}{
		{
			name:       "a single request shows",
			evs:        []events.Event{req("p1")},
			wantShown:  "p1",
			wantQueued: nil,
		},
		{
			name:       "a second request waits behind the first",
			evs:        []events.Event{req("p1"), req("p2")},
			wantShown:  "p1",
			wantQueued: []string{"p2"},
		},
		{
			// The bug that bit: answering the second closed the first
			// without answering it, stranding its turn.
			name:       "resolving the second leaves the first on screen",
			evs:        []events.Event{req("p1"), req("p2"), res("p2")},
			wantShown:  "p1",
			wantQueued: nil,
		},
		{
			name:       "resolving the shown one promotes the next",
			evs:        []events.Event{req("p1"), req("p2"), res("p1")},
			wantShown:  "p2",
			wantQueued: nil,
		},
		{
			name:       "resolving the shown one with nothing queued clears",
			evs:        []events.Event{req("p1"), res("p1")},
			wantShown:  "",
			wantQueued: nil,
		},
		{
			// Both halves of every permission live in the log, so a
			// reopened session replays each old answer in turn. Without
			// the drop, answering nothing fills the queue with answered
			// questions and the live modal sits behind them.
			name: "replay drops answered pairs and shows the live request",
			evs: []events.Event{
				req("p-old-1"), res("p-old-1"),
				req("p-old-2"), res("p-old-2"),
				req("p-live"),
			},
			wantShown:  "p-live",
			wantQueued: nil,
		},
		{
			name:       "a resolution for an unknown id changes nothing",
			evs:        []events.Event{req("p1"), req("p2"), res("p7")},
			wantShown:  "p1",
			wantQueued: []string{"p2"},
		},
		{
			name:       "three fanout requests drain in order",
			evs:        []events.Event{req("p1"), req("p2"), req("p3"), res("p1"), res("p2")},
			wantShown:  "p3",
			wantQueued: nil,
		},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			m := newTestModel()
			for _, ev := range tc.evs {
				m.applyEvent(ev)
			}
			var shown string
			if m.pending != nil {
				shown = m.pending.id
			}
			if shown != tc.wantShown {
				t.Errorf("on screen is %q, want %q", shown, tc.wantShown)
			}
			var queued []string
			for _, q := range m.pendingQueue {
				queued = append(queued, q.id)
			}
			if len(queued) != len(tc.wantQueued) {
				t.Fatalf("queued %v, want %v", queued, tc.wantQueued)
			}
			for i := range queued {
				if queued[i] != tc.wantQueued[i] {
					t.Errorf("queued %v, want %v", queued, tc.wantQueued)
					break
				}
			}
			// The composer is locked exactly while a request is on
			// screen: handleEnter refuses every message until the
			// queue empties, and sends once it has.
			m.input.SetValue("hello")
			_, cmd := handleEnter(m)
			if tc.wantShown == "" && cmd == nil {
				t.Error("Enter did nothing with the queue empty; the session is still wedged")
			}
			if tc.wantShown != "" && cmd != nil {
				t.Error("Enter sent a message while a request was still on screen")
			}
		})
	}
}

// Answering at the keyboard promotes the next queued request now, rather
// than waiting for this answer's own resolved event to come back — that
// event then matches nothing and is ignored, which is what an answer
// already given should do.
func TestAnsweringAtTheKeyboardPromotesTheNextRequest(t *testing.T) {
	m := armedPending(newTestModel(), false)
	m.pendingQueue = []*pendingPermission{
		{id: "p2", tool: "bash", description: "second", rule: "run *"},
	}

	next, cmd, handled := press(m, "y")
	if !handled || cmd == nil {
		t.Fatal("y did not answer the request on screen")
	}
	if next.pending == nil || next.pending.id != "p2" {
		t.Fatalf("the next request did not come up: %+v", next.pending)
	}
	if len(next.pendingQueue) != 0 {
		t.Fatalf("the promoted request stayed queued: %+v", next.pendingQueue)
	}
}
